src/lib/work_partitioner/job_pool.ml

```diff
@@ -1,42 +1,66 @@
 open Core_kernel
 
-type ('accum, 'final) fold_action =
-  | Continue of 'accum
-  | Continue_remove of 'accum
-  | Stop of 'final
-  | Stop_remove of 'final
-
-module Make (Id : Map.Key) (Spec : T) = struct
+module Make (Id : Hashtbl.Key) (Spec : T) = struct
   type job = (Spec.t, Id.t) Snark_work_lib.With_job_meta.t
 
-  module IdMap = Map.Make (Id)
-
-  type t = job IdMap.t
+  type t =
+    { timeline : Id.t Deque.t (* For iteration *)
+    ; index : (Id.t, job) Hashtbl.t (* For marking job as done *)
+    }
 
-  let first_job t : job option = IdMap.min_elt t |> Option.map ~f:Tuple2.get2
+  let create () =
+    { timeline = Deque.create (); index = Hashtbl.create (module Id) }
 
-  let fold_until ~init ~f ~finish t =
-    let seq = IdMap.to_sequence ~order:`Increasing_key t in
-    let handle_item (acc, retained) (k, v) =
-      match f acc v with
-      | Continue acc_new ->
-          Continue_or_stop.Continue (acc_new, retained)
-      | Continue_remove acc_new ->
-          Continue (acc_new, IdMap.remove retained k)
-      | Stop final ->
-          Stop (final, retained)
-      | Stop_remove final ->
-          Stop (final, IdMap.remove retained k)
-    in
+  let remove ~id t = Hashtbl.find_and_remove t.index id
 
-    Sequence.fold_until seq ~init:(init, t) ~f:handle_item
-      ~finish:(Tuple2.map_fst ~f:finish)
+  let find ~id t = Hashtbl.find t.index id
 
-  let add ~id ~job t = IdMap.add ~key:id ~data:job t
+  let change_inplace ~id ~f t = Hashtbl.change t.index id ~f
 
-  let change ~id ~f t = IdMap.change t id ~f
+  let rec remove_until_reschedule ~keep_condition ~should_reschedule t =
+    let%bind.Option job_id = Deque.dequeue_front t.timeline in
+    match Hashtbl.find t.index job_id with
+    | Some job when keep_condition job -> (
```

Comment:
I wonder, why not we make function accept `f`, a single function with type, instead of two different functions operating on `job`:

```ocaml
f :: job -> [ `Remove | `Keep | `Reschedule of job ]
```